### PR TITLE
Improve ddim scheduler and fix bug when prediction type is "sample"

### PR DESCRIPTION
**Pull request by PeterL1n**
This is a new PR to fix issue https://github.com/huggingface/diffusers/issues/1486.
The bug happens because when `prediction_type="sample"`, the computation for `pred_epsilon` was missing.
I also renamed the variable to make it more easy to reason about.

Closing the old PR: https://github.com/huggingface/diffusers/pull/1788

### CODE REVIEW

**Comment by PeterL1n** [comment]
@patrickvonplaten @patil-suraj Please take a look and approve.

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 1]
```diff
@@ -283,12 +283,13 @@ def step(
         # "predicted x_0" of formula (12) from https://arxiv.org/pdf/2010.02502.pdf
         if self.config.prediction_type == "epsilon":
             pred_original_sample = (sample - beta_prod_t ** (0.5) * model_output) / alpha_prod_t ** (0.5)
+            pred_epsilon = model_output
         elif self.config.prediction_type == "sample":
             pred_original_sample = model_output
+            pred_epsilon = (sample - alpha_prod_t ** (0.5) * pred_original_sample) / beta_prod_t ** (0.5)
```
Comment (PeterL1n) [inline src/diffusers/schedulers/scheduling_ddim.py:289]: This was missing in current code, causing the bug!

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for the PR. On first look, it makes sense. It would be nice if you could provide a code snippet which shows how this affects the results.

**Comment by PeterL1n** [comment]
I can't really provide a code snippet unless I have a model trained with prediction type "sample", which I currently don't at hand.

**Comment by patil-suraj** [comment]
Okay, I will try to train a model on a smaller dataset and then we could verify.

**Comment by clarencechen** [comment]
@patrickvonplaten @patil-suraj The [Diffusers Training Example Notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/training_example.ipynb) should provide a minimal example to compare DDIM sample results on models trained with different prediction types. I'll present a modified version of this notebook, including results with and without this PR, before the end of next week.

**Comment by patrickvonplaten** [comment]
Sorry for dropping the ball here - I'll prio this PR for early next week!

**Comment by clarencechen** [comment]
Here is a Github gist training a denoising diffusion model on the Huggingface sample butterfly dataset for 120 epochs using the linear noise schedule from DDPM, and then sampling images using both versions of DDIM with the same noise schedule:
https://colab.research.google.com/gist/clarencechen/ae750e54e59224a32f3bf2b1e23e60c3/diffusers_training_example.ipynb

**Comment by patrickvonplaten** [comment]
PR looks good thanks a lot for working on it!

**Pull request merged**